rfcs/45-stageless.md

```diff
@@ -0,0 +1,557 @@
+## User-facing explanation
+
+This explanation is, in effect, user-facing documentation for the new design.
+In addition to a few new concepts, it throws out much of the current system scheduling design that you may be familiar with.
+The following elements are substantially reworked:
+
+- schedules (flattened)
+- run criteria (can no longer loop, are now systems)
+- states (simplified, no longer purely run-criteria powered)
+- fixed time steps (no longer a run criteria)
+- exclusive systems (no longer special-cased)
+- command processing (now performed in a `flush_commands` exclusive system)
+- labels (can now be directly configured)
+- system sets (now just used to mass-apply a label to a group of systems)
+- stages (yeet)
+
+### Scheduling overview and intent-based configuration
+
+Systems in Bevy are stored in a `Schedule`: a collection of **configured systems**.
+Each frame, the `App`'s `runner` function will run the schedule,
+causing the **scheduler** to run the systems in parallel using a strategy that respects all of the configured constraints (or panic, if it is **unsatisfiable**).
+
+In the beginning, each `Schedule` is entirely unordered: systems will be selected in an arbitrary order and run if and only if all of the data that it must access is free.
+Just like with standard borrow checking, multiple systems can read from the same data at once, but writing to the data requires an exclusive lock.
+Systems which cannot be run in parallel are said to be **incompatible**.
+
+Bevy controls the ordering of its schedule(s) The solution via **intent-based configuration**: record the exact, minimal set of constraints needed for a scheduling strategy to be valid, and then let the scheduler figure it out!
+As the number of systems in your app grows, the number of possible scheduling strategies grows exponentially, and the set of *valid* scheduling strategies grows almost as quickly.
+However, each system will continue to work hand-in-hand with a small number of closely-related systems, interlocking in a straightforward and comprehensible way.
+
+By configuring our systems with straightforward, local rules, we can allow the scheduler to pick one of the possible valid paths, optimizing as it sees fit without breaking our logic.
+Just as importantly, we are not over-constraining our ordering, allowing subtle ordering bugs to surface quickly and then be stamped out with a well-considered rule.
+
+### The `App` stores multiple schedules
+
+The `App` can store multiple `Schedules` in a `HashMap<Box<dyn ScheduleLabel>, Schedule>` storage.
+This is used for the enter and exit schedules of states, but can also be used to store, mutate and access additional schedules.
+You can even mutate the schedule that you are in, although you cannot mutate systems that are currently running (failing with an error), as that would be UB.
+
+The main and startup schedules can be accessed using the `CoreSchedule::Main` and `CoreSchedule::Startup` labels respectively.

```

Comment:
The use of the term `Core` here suggests `bevy_core` to me, although these would live in `bevy_ecs` (I presume). This is probably fine though.

Reply:
I've been using `BaseSchedule`.